### DISCUSSION TITLE
Raspberry Pi compatibility

### DISCUSSION THREAD

**Discussion by KnorrFG** [hecrj/iced#1537] [General]
Hey, 
I am developing a project using iced that has to run on a Raspberry Pi. It used to work until recently (after mindlessly updating Cargo.lock), when suddenly It started giving me the following error, when running it on the Pi:

```
Error: the application graphics context could not be created
```

I tried switching to the glow renderer, which did not work, because it doesn't support Image.
I just noticed this on thursday,  and luckily on the same day, iced 0.5 was out, and with it a new glow renderer, which supports Image. 
So I ported the project over to 0.5, and had to notice that cross compilation didn't work anymore because of a new dependency:  
`servo-fontconfig-sys`.

However, I could still compile it on the raspi (which is of course way less comfortable than cross compiling), just to be greeted by the following error, when running the glow-version:
```
Xlib: sequence lost (0x10066 > 0x68) in reply type 0x0!
Xlib: sequence lost (0x1006b > 0x6d) in reply type 0x0!
Xlib: sequence lost (0x10070 > 0x72) in reply type 0x0!
Xlib: sequence lost (0x10075 > 0x77) in reply type 0x0!
Xlib: sequence lost (0x1007a > 0x7c) in reply type 0x0!
thread 'main' panicked at '0:23(17): error: no function with name 'texture'
', /home/neuro/.cargo/registry/src/github.com-1285ae84e5963aae/iced_glow-0.4.0/src/program.rs:77:17
```

While I can probably get the cross compilation to work somehow, I, so far, have not been able to figure out how to run my project on the Raspberry again (except for going back to the old lock file). I'm debating with myself to change the UI to gtk3, but that would be a lot of work. To be clear, I'm not blaming anyhone here. It's great that you develop such a library for free.

I guess I'm hoping for two things here: 
1. help to get it running on the Pi
2. starting a discussion about making cross-compilation and raspberry-pi compatibility an intended feature of iced.

Cheers Felix

---
**Edit:**

By now I found out, that installing the dependencies that are linked in fltk-rs' documentation will make the glow backend work on the raspi again. It might be a good Idea to document that somewhere. For reference, after installing these packages, it worked:
`libx11-dev libxext-dev libxft-dev libxinerama-dev libxcursor-dev libxrender-dev libxfixes-dev libpango1.0-dev libgl1-mesa-dev libglu1-mesa-dev`

**Comment by i509VCB** [comment]
Which raspberry pi do you have? The Pi 4 and Pi 3 have significantly different gpus from the perspective of glow and wgpu.

**Reply by KnorrFG** [reply]
Hey, it's a RaspberryPi 4 - Model B. 
By now I found out that also the wgpu renderer started working again. There were two RaspberryPis at play: one that I occasionally used for testing, but that was also used by colleagues, where I first noted that the program stopped working. It had the dependencies, which I mentioned in my edit, installed already. I then set up a second one to experiment. Without those dependencies installed, getting the same errors. I later installed them on the seccond raspberry pi, and it started working again. I just ran an upgrade and it started working again on the first Pi too. However, on both Pis (now with Iced 0.5) images would apear as black boxes, which used to work before all of this started. For both renderers, which is not suprising I guess, as both of them rely on glow. I then additionally installed `mesa-vulkan-drivers`, which then resulted in WGPU selecting the vulkan backend, and the images reappeared. 

I could also solve the cross compiling issues. I am using a docker container based on `debian:bullseye`, and all I had to do was adding those two lines to the Dockerfile:

```Docker
ENV PKG_CONFIG_LIBDIR_armv7_unknown_linux_gnueabihf=/usr/lib/arm-linux-gnueabihf/pkgconfig
ENV PKG_CONFIG_SYSROOT_DIR_armv7_unknown_linux_gnueabihf=/
```
This basically solves my first point, and only leaves the one about officially supporting Raspbian and cross-compilation.

**Reply by i509VCB** [reply]
The images in glow (does this occur with gles in wgpu?) seems odd.

**Reply by KnorrFG** [reply]
I don't know. I tried WGPU_BACKED=gles, it doesn't know that backend. I tried to figure out which backend uses gles by looking at <https://docs.rs/wgpu/latest/wgpu/enum.Backend.html>, and figured "maybe BrowserWebGpu", so I tried to use WGPU_BACKEND=browserwebgpu and WGPU_BACKEND=browser_webgpu, but neither was known.

**Reply by ids1024** [reply]
`WGPU_BACKEND=gl` should run it with OpenGL ES 3. Looks like the V3D graphics  in the Pi 4 should support that? Though beware that backend may not work as well as the wgpu Vulkan backend.

> Mesa’s V3D graphics driver stack includes a [conformant GLES3.1 driver](https://www.khronos.org/conformance/adopters/conformant-products/opengles#submission_882) called V3D and a Vulkan graphics driver called V3DV, notably used on the Raspberry Pi 4.

https://docs.mesa3d.org/drivers/v3d.html

**Reply by KnorrFG** [reply]
It does run with `WGPU_BACKEND=gl`, but then the images are black boxes.
